pkg/controller/core/resource_controller.go

```diff
@@ -54,10 +54,60 @@ const (
 	RequeueOnSuccess = 2 * time.Minute
 )
 
-// Commonly used reconciliation results.
+// Commonly used reconciliation results. Note that many of these results are
+// duplicates. They exist with different names in order to make the intentions
+// of the reconcile function explicit.
 var (
-	Result        = reconcile.Result{}
-	ResultRequeue = reconcile.Result{Requeue: true}
+	// DeprecatedResultDone indicates that no further reconcilations need to be
+	// queued for the object under reconciliation. It is deprecated and should
+	// be replaced with a more explicitly named reconcile result.
+	DeprecatedResultDone = reconcile.Result{Requeue: false}
+
+	// DeprecatedResultRequeue indicates that further reconciliations need to be
+	// queued for the object under reconciliattion. It is deprecated and should
+	// be replaced with a more explicitly named reconcile result.
+	DeprecatedResultRequeue = reconcile.Result{Requeue: true}
+
+	// ResultDirty should be returned when the reconciler has altered the object
+	// under reconciliation. When returning this result the reconciler must also
+	// submit the updated object to the API server, thus triggering an implicit
+	// reconciliation of the changed object.
+	ResultDirty = reconcile.Result{Requeue: false}
+
+	// ResultDeleted should be returned when the object under reconciliation has
+	// been deleted, i.e. after the reconciler has removed any finalizers.
+	ResultDeleted = ResultDirty
+
+	// ResultClean should be returned when the reconciler has not altered the
+	// object under reconciliation, and is not waiting for a pending operation
+	// to complete.
+	ResultClean = reconcile.Result{RequeueAfter: RequeueOnSuccess}
+
+	// ResultWaiting should be returned when the reconciler has not altered the
+	// object under reconciliation, but is waiting for a pending operation to
+	// complete, for example waiting for the managed cloud resource the object
+	// represents to transition from 'Creating' to 'Available'.
+	ResultWaiting = reconcile.Result{RequeueAfter: RequeueOnWait}
+
+	// ResultNotFound should be returned when the object under reconciliation no
+	// longer exists when the reconcile function is invoked.
+	ResultNotFound = reconcile.Result{Requeue: false}
+
+	// ResultError should be returned if the reconciler knows it will return an
+	// error, thus triggering an implicit reconciliation.
+	ResultError = reconcile.Result{Requeue: false}
```

Comment:
I am afraid, this may promote (or encorage) somewhat more verbose code. For example, instead of returning:
```go
return ResultWaiting, errors.Wrapf(r.Status().Update(ctx, intance), "failed to update status")
```
it will have to be written as:
```go
if err := r.Status().Update(ctx, instance); err != nil {
  return ResultError, errors.Wrapf(err, "failed to update status")
}
return ResultWait, nil
```

Reply:
I agree this would not be ideal. It's a different take on the same concept as the other conversation thread we have going so I'm going to resolve this one.